File: food/multistuff.py
import re
q = open('multistuff.settings.h', 'w')
s = ''
s += ('''#include "myjni.h"
#include <CoreFoundation/CoreFoundation.h>
#include "common.h"
#include <assert.h>
''')
s += ('''jboolean typeRefToBoolean(CFTypeRef val) {
    return (jboolean) CFBooleanGetValue(val);
}
CFTypeRef booleanToTypeRef(jboolean val) {
    return val ? kCFBooleanTrue : kCFBooleanFalse;
}
jobject typeRefToObject(CFTypeRef val) {
    return (void *) val;
}
CFTypeRef objectToTypeRef(jobject val) {
    return val;
}
''')

things = {
    'boolean': None,
    'object': None,
    'byte': 'kCFNumberSInt8Type',
    'char': 'kCFNumberSInt16Type',
    'short': 'kCFNumberSInt16Type',
    'int': 'kCFNumberLongType',
    'long': 'kCFNumberLongLongType',
    'float': 'kCFNumberFloatType',
    'double': 'kCFNumberDoubleType',
}

def impl(x, *alts):
    q.write('{\n')
    q.write('    void *x = stub(impl_%s);\n' % x)
    for a in [x] + list(alts):
        q.write('    env_functions.%s = x;\n' % a)
    q.write('}\n')

for jname, cfname in things.items():
    if cfname is not None:
        s += ('''j%s typeRefTo%s(CFTypeRef val) {
    j%s ret;
    _assert(CFNumberGetValue((CFNumberRef) val, %s, &ret));
    return ret;
}
''' % (jname, jname.capitalize(), jname, cfname))
        s += ('''CFTypeRef %sToTypeRef(j%s val) {
    return CFNumberCreate(NULL, %s, &val);
}
''' % (jname, jname, cfname))
    
    # GetXField
    # SetXField
    # GetStaticXField
    # SetStaticXField
    # CallXMethod[,V,A]
    # CallNonvirtualXMethod[,V,A]
    # CallStaticXMethod[,V,A]

    s += ('struct method_%s { j%s (*v)(jobject, va_list); j%s (*a)(jobject, const jvalue *); };\n' % (jname, jname, jname))

    s += ('''j%s impl_Call%sMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %%@", methodID); _abort(); }
    struct method_%s *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %%@", methodID); _abort(); }
    return st->a(obj, args);
}
''' % (jname, jname.capitalize(), jname))
    if jname == 'object':
        impl('CallObjectMethodA', 'CallNonvirtualObjectMethodA', 'CallStaticObjectMethodA', 'NewObjectA')
    else:
        impl('Call%sMethodA' % jname.capitalize(), 'CallNonvirtual%sMethodA' % jname.capitalize(), 'CallStatic%sMethodA' % jname.capitalize())

    s += ('''j%s impl_Call%sMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %%@", methodID); _abort(); }
    struct method_%s *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %%@", methodID); _abort(); }
    return st->v(obj, args);
}
''' % (jname, jname.capitalize(), jname))
    if jname == 'object':
        impl('CallObjectMethodV', 'CallNonvirtualObjectMethodV', 'CallStaticObjectMethodV', 'NewObjectV')
    else:
        impl('Call%sMethodV' % jname.capitalize(), 'CallNonvirtual%sMethodV' % jname.capitalize(), 'CallStatic%sMethodV' % jname.capitalize())

    s += ('''j%s impl_Call%sMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    j%s ret = impl_Call%sMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
''' % (jname, jname.capitalize(), jname, jname.capitalize()))
    if jname == 'object':
        impl('CallObjectMethod', 'CallNonvirtualObjectMethod', 'CallStaticObjectMethod', 'NewObject')
    else:
        impl('Call%sMethod' % jname.capitalize(), 'CallNonvirtual%sMethod' % jname.capitalize(), 'CallStatic%sMethod' % jname.capitalize())
    
    
    s += ('''void impl_Set%sField(JNIEnv *env, jobject obj, jfieldID fieldID, j%s val) {
    CFTypeRef ref = %sToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
''' % (jname.capitalize(), jname, jname))
    impl('Set%sField' % jname.capitalize(), 'SetStatic%sField' % jname.capitalize())
    
    s += ('''j%s impl_Get%sField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %%@", fieldID); _abort(); }
        return typeRefTo%s(ref);
}
''' % (jname, jname.capitalize(), jname.capitalize()))
    impl('Get%sField' % jname.capitalize(), 'GetStatic%sField' % jname.capitalize())

    if jname == 'object': continue
    



s += ('''
struct method_void { void (*v)(jobject, va_list); void (*a)(jobject, const     jvalue *); };
void impl_CallVoidMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const    jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef)          
methodID);
    if(!meth) { log("Unknown method (A): %%@", methodID); _abort(); }
    struct method_void *st; 
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %%@", methodID); _abort(); }
    st->a(obj, args);
}
void impl_CallVoidMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list  args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef)          
methodID);
    if(!meth) { log("Unknown method (V): %%@", methodID); _abort(); }
    struct method_void *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %%@", methodID); _abort(); }
    st->v(obj, args);
}
void impl_CallVoidMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    impl_CallVoidMethodV(env, obj, methodID, args);
    va_end(args);
}

''' % ())
impl('CallVoidMethod', 'CallNonvirtualVoidMethod', 'CallStaticVoidMethod')
impl('CallVoidMethodA', 'CallNonvirtualVoidMethodA', 'CallStaticVoidMethodA')
impl('CallVoidMethodV', 'CallNonvirtualVoidMethodV', 'CallStaticVoidMethodV')

f = open('multistuff.c', 'w')
f.write(s)
f.close()

g = open('multistuff.h', 'w')
for stuff in re.findall(re.compile('^.*{$', re.M), s):
    g.write(stuff[:-2] + ';\n')

File: food/multistuff.c
#include "myjni.h"
#include <CoreFoundation/CoreFoundation.h>
#include "common.h"
#include <assert.h>
jboolean typeRefToBoolean(CFTypeRef val) {
    return (jboolean) CFBooleanGetValue(val);
}
CFTypeRef booleanToTypeRef(jboolean val) {
    return val ? kCFBooleanTrue : kCFBooleanFalse;
}
jobject typeRefToObject(CFTypeRef val) {
    return (void *) val;
}
CFTypeRef objectToTypeRef(jobject val) {
    return val;
}
jchar typeRefToChar(CFTypeRef val) {
    jchar ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberSInt16Type, &ret));
    return ret;
}
CFTypeRef charToTypeRef(jchar val) {
    return CFNumberCreate(NULL, kCFNumberSInt16Type, &val);
}
struct method_char { jchar (*v)(jobject, va_list); jchar (*a)(jobject, const jvalue *); };
jchar impl_CallCharMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_char *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jchar impl_CallCharMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_char *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jchar impl_CallCharMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jchar ret = impl_CallCharMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetCharField(JNIEnv *env, jobject obj, jfieldID fieldID, jchar val) {
    CFTypeRef ref = charToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jchar impl_GetCharField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToChar(ref);
}
jfloat typeRefToFloat(CFTypeRef val) {
    jfloat ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberFloatType, &ret));
    return ret;
}
CFTypeRef floatToTypeRef(jfloat val) {
    return CFNumberCreate(NULL, kCFNumberFloatType, &val);
}
struct method_float { jfloat (*v)(jobject, va_list); jfloat (*a)(jobject, const jvalue *); };
jfloat impl_CallFloatMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_float *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jfloat impl_CallFloatMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_float *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jfloat impl_CallFloatMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jfloat ret = impl_CallFloatMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetFloatField(JNIEnv *env, jobject obj, jfieldID fieldID, jfloat val) {
    CFTypeRef ref = floatToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jfloat impl_GetFloatField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToFloat(ref);
}
struct method_boolean { jboolean (*v)(jobject, va_list); jboolean (*a)(jobject, const jvalue *); };
jboolean impl_CallBooleanMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_boolean *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jboolean impl_CallBooleanMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_boolean *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jboolean impl_CallBooleanMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jboolean ret = impl_CallBooleanMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetBooleanField(JNIEnv *env, jobject obj, jfieldID fieldID, jboolean val) {
    CFTypeRef ref = booleanToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jboolean impl_GetBooleanField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToBoolean(ref);
}
jint typeRefToInt(CFTypeRef val) {
    jint ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberLongType, &ret));
    return ret;
}
CFTypeRef intToTypeRef(jint val) {
    return CFNumberCreate(NULL, kCFNumberLongType, &val);
}
struct method_int { jint (*v)(jobject, va_list); jint (*a)(jobject, const jvalue *); };
jint impl_CallIntMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_int *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jint impl_CallIntMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_int *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jint impl_CallIntMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jint ret = impl_CallIntMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetIntField(JNIEnv *env, jobject obj, jfieldID fieldID, jint val) {
    CFTypeRef ref = intToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jint impl_GetIntField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToInt(ref);
}
jdouble typeRefToDouble(CFTypeRef val) {
    jdouble ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberDoubleType, &ret));
    return ret;
}
CFTypeRef doubleToTypeRef(jdouble val) {
    return CFNumberCreate(NULL, kCFNumberDoubleType, &val);
}
struct method_double { jdouble (*v)(jobject, va_list); jdouble (*a)(jobject, const jvalue *); };
jdouble impl_CallDoubleMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_double *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jdouble impl_CallDoubleMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_double *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jdouble impl_CallDoubleMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jdouble ret = impl_CallDoubleMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetDoubleField(JNIEnv *env, jobject obj, jfieldID fieldID, jdouble val) {
    CFTypeRef ref = doubleToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jdouble impl_GetDoubleField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToDouble(ref);
}
jbyte typeRefToByte(CFTypeRef val) {
    jbyte ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberSInt8Type, &ret));
    return ret;
}
CFTypeRef byteToTypeRef(jbyte val) {
    return CFNumberCreate(NULL, kCFNumberSInt8Type, &val);
}
struct method_byte { jbyte (*v)(jobject, va_list); jbyte (*a)(jobject, const jvalue *); };
jbyte impl_CallByteMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_byte *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jbyte impl_CallByteMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_byte *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jbyte impl_CallByteMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jbyte ret = impl_CallByteMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetByteField(JNIEnv *env, jobject obj, jfieldID fieldID, jbyte val) {
    CFTypeRef ref = byteToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jbyte impl_GetByteField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToByte(ref);
}
struct method_object { jobject (*v)(jobject, va_list); jobject (*a)(jobject, const jvalue *); };
jobject impl_CallObjectMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_object *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jobject impl_CallObjectMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_object *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jobject impl_CallObjectMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jobject ret = impl_CallObjectMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetObjectField(JNIEnv *env, jobject obj, jfieldID fieldID, jobject val) {
    CFTypeRef ref = objectToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jobject impl_GetObjectField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToObject(ref);
}
jlong typeRefToLong(CFTypeRef val) {
    jlong ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberLongLongType, &ret));
    return ret;
}
CFTypeRef longToTypeRef(jlong val) {
    return CFNumberCreate(NULL, kCFNumberLongLongType, &val);
}
struct method_long { jlong (*v)(jobject, va_list); jlong (*a)(jobject, const jvalue *); };
jlong impl_CallLongMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_long *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jlong impl_CallLongMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_long *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jlong impl_CallLongMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jlong ret = impl_CallLongMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetLongField(JNIEnv *env, jobject obj, jfieldID fieldID, jlong val) {
    CFTypeRef ref = longToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jlong impl_GetLongField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToLong(ref);
}
jshort typeRefToShort(CFTypeRef val) {
    jshort ret;
    _assert(CFNumberGetValue((CFNumberRef) val, kCFNumberSInt16Type, &ret));
    return ret;
}
CFTypeRef shortToTypeRef(jshort val) {
    return CFNumberCreate(NULL, kCFNumberSInt16Type, &val);
}
struct method_short { jshort (*v)(jobject, va_list); jshort (*a)(jobject, const jvalue *); };
jshort impl_CallShortMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_short *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    return st->a(obj, args);
}
jshort impl_CallShortMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef) methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_short *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    return st->v(obj, args);
}
jshort impl_CallShortMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    jshort ret = impl_CallShortMethodV(env, obj, methodID, args);
    va_end(args);
    return ret;
}
void impl_SetShortField(JNIEnv *env, jobject obj, jfieldID fieldID, jshort val) {
    CFTypeRef ref = shortToTypeRef(val);
    CFDictionarySetValue(obj, (CFStringRef) fieldID, ref);
    CFRelease(ref);
}
jshort impl_GetShortField(JNIEnv *env, jobject obj, jfieldID fieldID) {
        CFTypeRef ref = CFDictionaryGetValue(obj, (CFStringRef) fieldID);
        if(!ref) { log("Unknown property: %@", fieldID); _abort(); }
        return typeRefToShort(ref);
}

struct method_void { void (*v)(jobject, va_list); void (*a)(jobject, const     jvalue *); };
void impl_CallVoidMethodA(JNIEnv *env, jobject obj, jmethodID methodID, const    jvalue *args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef)          
methodID);
    if(!meth) { log("Unknown method (A): %@", methodID); _abort(); }
    struct method_void *st; 
    _assert(st = RawDataGetPtr(meth));
    if(!st->a) { log("No A impl: %@", methodID); _abort(); }
    st->a(obj, args);
}
void impl_CallVoidMethodV(JNIEnv *env, jobject obj, jmethodID methodID, va_list  args) {
    const void *meth = CFDictionaryGetValue(obj, (CFStringRef)          
methodID);
    if(!meth) { log("Unknown method (V): %@", methodID); _abort(); }
    struct method_void *st;
    _assert(st = RawDataGetPtr(meth));
    if(!st->v) { log("No V impl: %@", methodID); _abort(); }
    st->v(obj, args);
}
void impl_CallVoidMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...) {
    va_list args;
    va_start(args, methodID);
    impl_CallVoidMethodV(env, obj, methodID, args);
    va_end(args);
}

